src/lib/util/hash.cpp

```diff
@@ -27,48 +30,80 @@ char const *const hash_collection::HASH_TYPES_CRC_SHA1 = "RS";
 char const *const hash_collection::HASH_TYPES_ALL = "RS";
 
 
-
 //**************************************************************************
-//  HASH COLLECTION
+//  HASH CREATOR
 //**************************************************************************
 
-//-------------------------------------------------
-//  hash_collection - constructor
-//-------------------------------------------------
+namespace {
 
-hash_collection::hash_collection()
-	: m_has_crc32(false),
-		m_has_sha1(false),
-		m_creator(nullptr)
+class hash_creator : public write_stream
 {
-}
+public:
+	// constructor
+	hash_creator(hash_collection &hashes, bool doing_crc32, bool doing_sha1)
+		: m_hashes(hashes)
+	{
+		if (doing_crc32)
+			m_crc32_creator = std::make_optional<crc32_creator>();
+		if (doing_sha1)
+			m_sha1_creator = std::make_optional<sha1_creator>();
+	}
+
+	// add the given buffer to the hash
+	virtual std::error_condition write(void const *buffer, std::size_t length, std::size_t &actual) noexcept override
+	{
+		// append to each active hash
+		if (m_crc32_creator.has_value())
+			m_crc32_creator.value().append(buffer, length);
+		if (m_sha1_creator.has_value())
+			m_sha1_creator.value().append(buffer, length);
+
+		actual = length;
+		return std::error_condition();
+	}
 
+	// dummy override
+	virtual std::error_condition flush() noexcept override
+	{
+		return std::error_condition();
+	}
 
-hash_collection::hash_collection(std::string_view string)
-	: m_has_crc32(false),
-		m_has_sha1(false),
-		m_creator(nullptr)
-{
-	from_internal_string(string);
-}
+	// stop hashing
+	virtual std::error_condition finalize() noexcept override
```

Comment:
There are multiple pitfalls with this:
* This stream requires explicit finalisation.  Generally, it’s expected that write streams will finalise themselves on destruction if it wasn’t done explicitly (e.g. see the `zlib_write_filter`).
* It doesn’t either reset or detach itself on finalisation.  It should either reset itself so that it can be used to digest another batch of data, or it should detach itself and return an error on future writes.

Reply:
I'm not sure that finalize-on-destruction is a better semantic than explicit finalization. I think the option to not finalize a stream if something goes wrong during the process of writing to it could be worthwhile.

The "reset or detach" semantics don't seem to be documented anywhere.